src/gui/guiScene.cpp

```diff
@@ -0,0 +1,262 @@
+		rotateCamera(v3f(m_custom_rot.X, m_custom_rot.Y, 0.f));
+		calcOptimalDistance();
+
+		m_initial_rotation = false;
+	}
+
+	m_driver->setViewPort(oldViewPort);
+}
+
+bool GUIScene::OnEvent(const SEvent &event)
+{
+	if (m_mouse_ctrl) {
+		switch(event.EventType) {
+		case EET_MOUSE_INPUT_EVENT:
+			if (event.MouseInput.Event == EMIE_LMOUSE_PRESSED_DOWN) {
+				m_last_pos = v2f((f32)event.MouseInput.X, (f32)event.MouseInput.Y);
+				return true;
+			} else if (event.MouseInput.Event == EMIE_MOUSE_MOVED) {
+				if (event.MouseInput.isLeftPressed()) {
+					m_curr_pos = v2f((f32)event.MouseInput.X, (f32)event.MouseInput.Y);
+
+					rotateCamera(v3f(
+						m_last_pos.Y - m_curr_pos.Y,
+						m_curr_pos.X - m_last_pos.X, 0.f));
+
+					m_last_pos = m_curr_pos;
+					return true;
+				}
+			}
+			break;
+		default: break;
+		}
+	}
+
+	return gui::IGUIElement::OnEvent(event);
+}
+
+void GUIScene::setStyles(const std::array<StyleSpec, StyleSpec::NUM_STATES> &styles)
+{
+	StyleSpec::State state = StyleSpec::STATE_DEFAULT;
+	StyleSpec style = StyleSpec::getStyleFromStatePropagation(styles, state);
+
+	setNotClipped(style.getBool(StyleSpec::NOCLIP, false));
+	setBackgroundColor(style.getColor(StyleSpec::BGCOLOR, m_bgcolor));
+}
+
+/* Camera control functions */
+
+inline void GUIScene::calcOptimalDistance()
+{
+	core::aabbox3df box = m_mesh->getBoundingBox();
+	f32 width  = box.MaxEdge.X - box.MinEdge.X;
+	f32 height = box.MaxEdge.Y - box.MinEdge.Y;
+	f32 depth  = box.MaxEdge.Z - box.MinEdge.Z;
+	f32 max_width = width > depth ? width : depth;
+
+	const scene::SViewFrustum *f = m_cam->getViewFrustum();
+	f32 cam_far = m_cam->getFarValue();
+	f32 far_width = core::line3df(f->getFarLeftUp(), f->getFarRightUp()).getLength();
+	f32 far_height = core::line3df(f->getFarLeftUp(), f->getFarLeftDown()).getLength();
+
+	core::recti rect = getAbsolutePosition();
+	f32 zoomX = rect.getWidth() / max_width;
+	f32 zoomY = rect.getHeight() / height;
+	f32 dist;
+
+	if (zoomX < zoomY)
+		dist = (max_width / (far_width / cam_far)) + (0.5f * max_width);
+	else
+		dist = (height / (far_height / cam_far)) + (0.5f * max_width);
+
+	m_cam_distance = dist;
+	m_update_cam = true;
+}
+
+void GUIScene::updateCamera(scene::ISceneNode *target)
+{
+	m_target = target;
+	updateTargetPos();
+
+	m_last_target_pos = m_target_pos;
+	updateCameraPos();
+
+	m_update_cam = true;
+}
+
+void GUIScene::updateTargetPos()
+{
+	m_last_target_pos = m_target_pos;
+	m_target->updateAbsolutePosition();
+	m_target_pos = m_target->getAbsolutePosition();
+}
+
+void GUIScene::setCameraRotation(v3f rot)
+{
+	correctBounds(rot);
+
+	core::matrix4 mat;
+	mat.setRotationDegrees(rot);
+
+	m_cam_pos = v3f(0.f, 0.f, m_cam_distance);
+	mat.rotateVect(m_cam_pos);
+
+	m_cam_pos += m_target_pos;
+	m_cam->setPosition(m_cam_pos);
+	m_update_cam = false;
+}
+
+bool GUIScene::correctBounds(v3f &rot)
+{
+	static const float ROTATION_MAX_1 = 60.0f;
+	static const float ROTATION_MAX_2 = 300.0f;
```

Comment:
I don't know much about it, but I think a `constexpr` would be more appropriate.